File: leetcode/C/206.反转链表.c
/*
 * @lc app=leetcode.cn id=206 lang=c
 *
 * [206] 反转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseList(struct ListNode* head) {
   /*翻转指针
    //判断链表是否为空
    if(head == NULL)
    {
        return NULL;
    }
    
    struct ListNode* n1,*n2,*n3;
    n1=NULL;
    n2=head;
    n3=head->next;

    while(n2)
    {
        //翻转
        n2->next = n1;

        //迭代往后走
        n1=n2;
        n2=n3;
        if(n3)
        {
        n3=n3->next;
        }
    }
    return n1;
    */

   /*头插：取原链表中节点，头插到newhead新链表中*/
   struct ListNode*cur =head;
   struct ListNode*newhead = NULL;

   while(cur)
   {
    struct ListNode*next = cur->next;

    //头插
    cur->next = newhead;
    newhead = cur;

    //迭代往后走
    cur = next;
   }

   return newhead;

}
// @lc code=end

File: leetcode/Java/206.反转链表.java
/*
 * @lc app=leetcode.cn id=206 lang=java
 *
 * [206] 反转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

// 双指针法
class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        ListNode cur = head;
        ListNode temp = null;

        while (cur != null) {
            temp = cur.next; // 保存下一个节点
            cur.next = prev; // 翻转指针
            prev = cur;
            cur =temp;
        }
        return prev;
    }
}

// 递归 
class Solution {
    public ListNode reverseList(ListNode head) {
        return reverse(null, head);
    }

    private ListNode reverse(ListNode prev, ListNode cur) {
        if (cur == null) {
            return prev;
        }
        ListNode temp = null;
        temp = cur.next;// 先保存下一个节点
        cur.next = prev;// 反转
        // 更新prev、cur位置
        // prev = cur;
        // cur = temp;
        return reverse(cur, temp);
    }
}

// @lc code=end

